### ISSUE TITLE
Replace 'data' with 'prior(this)'

### ISSUE THREAD

**Issue by mckoss** [firebase/bolt#35]
It's odd to have a "global" symbol represent the prior state of a data element.

I propose instead to offer a built-in-function, prior(), that we return the previous value of a datum (one referenced via this).

    'prior(this.prop)' => data.prop

Note how you can then write:

    function nonDecreasing(n) { return prior(n) == null || n > prior(n); }
    path /x { write() { return nonDecreasing(this.counter);  }

@tomlarkworthy @matjaz @TristonianJones - Comments?

**Comment by mckoss** [comment]
If run in the context of a read, or a non-changed value as part of a write prior(this) == this.

**Comment by matjaz** [comment]
It's longer than just typing `data` :]

I guess I can call `nonDecreasing()` from type?
```
type NonDecreasing extends Number {
  validate() { return this > prior(this); }
}
```

This relates to #33.

**Comment by tomlarkworthy** [comment]
I like the idea. FWIW in temporal logic of actions (TLA) and other temporal logic systems, they usually accentuate the next value rather than the previous value
i.e.
x' = x + 1
next(x) = x + 1

given .read evaluates on the NOW, I feel like writes evaluate on the NOW + FUTURE. So the TLA style is slightly more consistent, in contrast to going to the PAST.

**Comment by matjaz** [comment]
By the definition `this` is future value (`x'`). I wouldn't change this.

**Comment by tomlarkworthy** [comment]
Yeah it's not important. Just drawing attention to all the things I can think of for due diligence. I like a temporal function that can generalize to prior(root) as well better than a set of globals that we have today.

**Comment by mckoss** [comment]
Fixed by 3a5b4bb3d7e6b357e71a3b8f4c77ad16591b4686
